icicle-compiler/test/Icicle/Test/Sea/Zebra.hs

```diff
@@ -0,0 +1,374 @@
+        _ ->
+          Nothing
+
+    --fromVArray f val =
+    --  case val of
+    --    VArray vs ->
+    --      f vs
+    --    _ ->
+    --      Nothing
+
+    fromVOption f val =
+      case val of
+        VNone ->
+          (0,) <$> f []
+        VSome v ->
+          (1,) <$> f [v]
+        _ ->
+          Nothing
+
+    fromVSumError f val =
+      case val of
+        VLeft (VError e) ->
+          (fromIntegral (wordOfError e),) <$> f []
+        VRight v ->
+          (fromIntegral (wordOfError ExceptNotAnError),) <$> f [v]
+        _ ->
+          Nothing
+
+    fromVSum f g val =
+      case val of
+        VLeft v ->
+          (0,,) <$> f [v] <*> g []
+        VRight v ->
+          (1,,) <$> f [] <*> g [v]
+        _ ->
+          Nothing
+
+    fromVError val =
+      case val of
+        VError _ ->
+          pure 1
+        _ ->
+          Nothing
+
+    mergeValues t vs
+      | null vs
+      = tableOf t []
+      | otherwise
+      = pure (concat vs)
+
+    tableOf :: ValType -> [BaseValue] -> Maybe [Zebra.Column ()]
+    tableOf ty rows =
+      case ty of
+        BoolT ->
+          pure . Zebra.IntColumn . Storable.fromList <$> mapM fromVBool rows
+
+        TimeT ->
+          pure . Zebra.IntColumn . Storable.fromList <$> mapM fromVTime rows
+
+        DoubleT ->
+          pure . Zebra.DoubleColumn . Storable.fromList <$> mapM fromVDouble rows
+
+        IntT ->
+          pure . Zebra.IntColumn . Storable.fromList <$> mapM fromVInt rows
+
+        UnitT ->
+          pure . Zebra.IntColumn . Storable.fromList <$> mapM fromVUnit rows
+
+        ErrorT ->
+          pure . Zebra.IntColumn . Storable.fromList <$> mapM fromVError rows
+
+        StringT -> do
+          strings <- mapM fromVString rows
+
+          let
+            string =
+              ByteString.concat strings
+            len =
+              ByteString.length string
+            lens =
+              fmap (fromIntegral . ByteString.length) strings
+
+          pure $ [ Zebra.ArrayColumn
+                     (Storable.fromList lens)
+                     (Zebra.Table () len (Boxed.fromList [ Zebra.ByteColumn string ]))
+                 ]
+
+        -- FIXME: we aren't reading nested arrays right now (savages)
+        --ArrayT t -> do
+        --  vs <- concat <$> mapM (fromVArray (tableOf t)) rows
+        --  lens <- mapM (fromVArray (pure . fromIntegral . length)) rows
+        --  pure [ Zebra.ArrayColumn
+        --           (Storable.fromList lens)
+        --           (Zebra.Table () (length vs) (Boxed.fromList vs))
+        --       ]
+
+        OptionT t -> do
+          (bools, vs) <- List.unzip <$> mapM (fromVOption (tableOf t)) rows
+          cols <- mergeValues t vs
+          pure $ [ Zebra.IntColumn (Storable.fromList bools) ] <> cols
+
+        SumT ErrorT t -> do
+          (bools, vs) <- List.unzip <$> mapM (fromVSumError (tableOf t)) rows
+          pure $ [ Zebra.IntColumn (Storable.fromList bools) ] <> concat vs
```

Comment:
this is how icicle does it, but not sure that this specialisation should end up in a zebra file

should probably be `(tags, errs, vs)` instead?